File: haskell/heap.hs
-- Binary heap maintenance under a Gaussian workload — purely functional
-- form: a Braun min-heap (size-balanced tree; the idiomatic persistent
-- binary heap), same Irwin-Hall stream and unconditional replace-top
-- maintenance as the array variants. The checksum only depends on the heap
-- *contents* (evicted minima + final multiset, mod 1e9+7), which is
-- arrangement-independent, so it matches the array implementations exactly.
{-# LANGUAGE BangPatterns #-}

module Main where

import System.Exit (exitFailure)
import Data.Int (Int64)

data Heap = Leaf | Node !Int64 !Heap !Heap

p :: Int64
p = 1000000007

lcg :: Int64 -> Int64
lcg x = (x * 48271) `rem` 2147483647

gauss :: Int64 -> (Int64, Int64)
gauss = go 12 0
  where
    go 0 !s !x = (s, x)
    go !k !s !x = let x1 = lcg x in go (k - 1) (s + x1) x1

insertH :: Int64 -> Heap -> Heap
insertH v Leaf = Node v Leaf Leaf
insertH v (Node w l r)
  | v <= w = Node v (insertH w r) l
  | otherwise = Node w (insertH v r) l

-- Replace the root value and restore the heap property downward.
replaceTop :: Int64 -> Heap -> Heap
replaceTop v (Node _ l r) = down v l r
replaceTop v Leaf = Node v Leaf Leaf

down :: Int64 -> Heap -> Heap -> Heap
down !v l r =
  case (l, r) of
    (Node lv ll lr, Node rv rl rr)
      | lv <= rv ->
          if v <= lv then Node v l r else Node lv (down v ll lr) r
      | otherwise ->
          if v <= rv then Node v l r else Node rv l (down v rl rr)
    (Node lv ll lr, Leaf) ->
      if v <= lv then Node v l r else Node lv (down v ll lr) Leaf
    (Leaf, Node rv rl rr) ->
      if v <= rv then Node v l r else Node rv Leaf (down v rl rr)
    (Leaf, Leaf) -> Node v Leaf Leaf

top :: Heap -> Int64
top (Node w _ _) = w
top Leaf = 0

sumH :: Heap -> Int64 -> Int64
sumH Leaf !acc = acc
sumH (Node w l r) !acc = sumH r (sumH l ((acc + w) `rem` p))

heapTest :: Int -> Int64
heapTest m0 = go m0 x0 h0 0
  where
    (h0, x0) = build (0 :: Int) 20260711 Leaf
    build !k !x !h
      | k == 65535 = (h, x)
      | otherwise = let (g, x') = gauss x in build (k + 1) x' (insertH g h)
    go 0 _ !h !acc = (acc + sumH h 0) `rem` p
    go !k !x !h !acc =
      let (g, x') = gauss x
          !t = top h
       in go (k - 1) x' (replaceTop (t + g) h) ((acc + t) `rem` p)

main :: IO ()
main = do
  let c = heapTest 6500000
  if c /= 558972311
    then putStrLn ("FAIL: expected 558972311, got " ++ show c) >> exitFailure
    else pure ()

File: rust/heap.rs
// Binary heap maintenance under a Gaussian workload — direct-mutation
// baseline: min-heap of 65535 i64 in a Vec, Irwin-Hall draws (sum of 12
// consecutive MINSTD outputs), M rounds of unconditional replace-top.
// Checksum = (sum of evicted minima + final heap contents) mod 1e9+7.

const N: usize = 65535;
const P: i64 = 1000000007;

fn lcg(x: i64) -> i64 {
    (x * 48271) % 2147483647
}

fn gauss(x: i64) -> (i64, i64) {
    let mut s = 0i64;
    let mut x = x;
    for _ in 0..12 {
        x = lcg(x);
        s += x;
    }
    (s, x)
}

fn sift_up(a: &mut [i64], mut i: usize) {
    while i > 0 {
        let p = (i - 1) / 2;
        if a[i] < a[p] {
            a.swap(i, p);
            i = p;
        } else {
            break;
        }
    }
}

fn sift_down(a: &mut [i64], mut i: usize) {
    loop {
        let l = 2 * i + 1;
        if l >= N {
            break;
        }
        let r = l + 1;
        let c = if r < N && a[r] < a[l] { r } else { l };
        if a[c] < a[i] {
            a.swap(i, c);
            i = c;
        } else {
            break;
        }
    }
}

fn heap_test(m: i64) -> i64 {
    let mut a = vec![0i64; N];
    let mut x = 20260711i64;
    for k in 0..N {
        let (g, x1) = gauss(x);
        x = x1;
        a[k] = g;
        sift_up(&mut a, k);
    }
    let mut acc = 0i64;
    for _ in 0..m {
        let (g, x1) = gauss(x);
        x = x1;
        let top = a[0];
        acc = (acc + top) % P;
        a[0] = top + g;
        sift_down(&mut a, 0);
    }
    let mut s = 0i64;
    for &v in &a {
        s = (s + v) % P;
    }
    (acc + s) % P
}

fn main() {
    let m: i64 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(26000000);
    println!("{}", heap_test(m));
}
